File: modules/data/trand.lua
sptbl["trand"] = {

    files = {
        module = "trand.c",
        header = "trand.h",
        example = "ex_trand.c",
    },

    func = {
        create = "ut_trand_create",
        destroy = "ut_trand_destroy",
        init = "ut_trand_init",
        compute = "ut_trand_compute",
    },

    params = {
        optional = {
            {
                name = "min",
                type = "UTFLOAT",
                description ="Minimum value to use.",
                default = 0
            },
            {
                name = "max",
                type = "UTFLOAT",
                description ="Maximum value to use.",
                default = 1
            },
        }
    },

    modtype = "module",

    description = [[Triggered random number generator.
]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "trig",
            description = "When non-zero, will create a new random value."
        },
    },

    outputs = {
        {
            name = "out",
            description = "Signal output."
        },
    }

}

File: modules/trand.c
#include <stdlib.h>
#include "utone.h"

int ut_trand_create(ut_trand **p)
{
    *p = malloc(sizeof(ut_trand));
    return UT_OK;
}

int ut_trand_destroy(ut_trand **p)
{
    free(*p);
    return UT_OK;
}

int ut_trand_init(ut_data *ut, ut_trand *p)
{
    p->min = 0;
    p->max = 1;
    p->val = 0;
    return UT_OK;
}

int ut_trand_compute(ut_data *ut, ut_trand *p, UTFLOAT *in, UTFLOAT *out)
{
    if(*in != 0) {
        p->val = p->min + ((UTFLOAT) ut_rand(ut) / UT_RANDMAX) * (p->max - p->min);
        *out = p->val;
    } else {
        *out = p->val;
    }
    return UT_OK;
}
